Sources/NIOPosix/Bootstrap.swift

```diff
@@ -1352,6 +1352,482 @@ public final class ClientBootstrap: NIOClientTCPBootstrapProtocol {
+                .flatMap {
+                    if let bindTarget = bindTarget {
+                        return channel
+                            .bind(to: bindTarget)
+                            .flatMap {
+                                channelInitializer(channel)
+                            }.flatMapThrowing {
+                                return try NIOAsyncChannel(
+                                    synchronouslyWrapping: channel,
+                                    backpressureStrategy: channelBackpressureStrategy,
+                                    isOutboundHalfClosureEnabled: isOutboundHalfClosureEnabled,
+                                    inboundType: channelInboundType,
+                                    outboundType: channelOutboundType
+                                )
+                            }
+                    } else {
+                        return channelInitializer(channel)
+                            .flatMapThrowing {
+                                return try NIOAsyncChannel(
+                                    synchronouslyWrapping: channel,
+                                    backpressureStrategy: channelBackpressureStrategy,
+                                    isOutboundHalfClosureEnabled: isOutboundHalfClosureEnabled,
+                                    inboundType: channelInboundType,
+                                    outboundType: channelOutboundType
+                                )
+                            }
+                    }
+                }.flatMap { (asyncChannel: NIOAsyncChannel<ChildChannelInbound, ChildChannelOutbound>) in
+                    eventLoop.assertInEventLoop()
+                    return channel.registerAndDoSynchronously(body).map { asyncChannel }
+                }.flatMapError { error in
+                    channel.close0(error: error, mode: .all, promise: nil)
+                    return channel.eventLoop.makeFailedFuture(error)
+                }
+        }
+
+        if eventLoop.inEventLoop {
+            return setupChannel()
+        } else {
+            return eventLoop.flatSubmit {
+                setupChannel()
+            }
+        }
+    }
+}
+
+// MARK: AsyncChannel based connect with protocol negotiation
+extension ClientBootstrap {
+    /// Specify the `host` and `port` to connect to for the TCP `Channel` that will be established.
+    ///
+    /// - Parameters:
+    ///   - host: The host to connect to.
+    ///   - port: The port to connect to.
+    ///   - channelInitializer: A closure to initialize the channel which must return the handler that is used for negotiating
+    ///   the protocol.
+    /// - Returns: The protocol negotiation result.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func connect<Handler: NIOProtocolNegotiationHandler>(
+        host: String,
+        port: Int,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Handler>
```

Comment:
This PR uses a different approach than the `ServerBootstrap` where we passed the type of the protocol negotiation handler. I am going to check and see if we can make this approach work for `ServerBootstrap` as well